libsolidity/interface/CompilerStack.cpp

```diff
@@ -1464,25 +1464,36 @@ void CompilerStack::generateIR(ContractDefinition const& _contract)
 		);
 	}
 
-	yul::YulStack stack(
-		m_evmVersion,
-		m_eofVersion,
-		yul::YulStack::Language::StrictAssembly,
-		m_optimiserSettings,
-		m_debugInfoSelection
-	);
-	bool yulAnalysisSuccessful = stack.parseAndAnalyze("", compiledContract.yulIR);
-	solAssert(
-		yulAnalysisSuccessful,
-		compiledContract.yulIR + "\n\n"
-		"Invalid IR generated:\n" +
-		langutil::SourceReferenceFormatter::formatErrorInformation(stack.errors(), stack) + "\n"
-	);
+	auto const reparseYul = [&](std::string const& _irSource) {
+		YulStack stack(
+			m_evmVersion,
+			m_eofVersion,
+			YulStack::Language::StrictAssembly,
+			m_optimiserSettings,
+			m_debugInfoSelection
+		);
+		bool yulAnalysisSuccessful = stack.parseAndAnalyze("", _irSource);
+		solAssert(
+			yulAnalysisSuccessful,
+			_irSource + "\n\n"
+			"Invalid IR generated:\n" +
+			langutil::SourceReferenceFormatter::formatErrorInformation(stack.errors(), stack) + "\n"
+		);
+		return stack;
+	};
 
-	compiledContract.yulIRAst = stack.astJson();
-	stack.optimize();
-	compiledContract.yulIROptimized = stack.print(this);
-	compiledContract.yulIROptimizedAst = stack.astJson();
+	{
+		YulStack stack = reparseYul(compiledContract.yulIR);
+		compiledContract.yulIRAst = stack.astJson();
+		stack.optimize();
+		compiledContract.yulIROptimized = stack.print(this);
+	}
+	{
+		// Optimizer does not maintain correct native source locations in the AST.
+		// We can work around it by regenerating the AST from scratch from optimized IR.
+		YulStack stack = reparseYul(compiledContract.yulIROptimized);
+		compiledContract.yulIROptimizedAst = stack.astJson();
```

Comment:
So just to fully understand; before:

1. parse generated IR (unoptimized) - this generates native source locations
2. optimize (this changes the IR without updating the native source locations, hence optimized IR pointing to location in unoptimized IR)
3. assign optimized IR AST to contract (with wrong source locations)

now:

1. after optimizing, parse optimized yul IR again (this updates the native source locations to reflect the optimized IR instead of unoptimized)
2. assign said IR (optimized) AST to contract

Correct?

Reply:
Yeah pretty much.

Just note that what I'm assigning is the JSON-serialization of AST, not the original AST (we discard that along with `YulStack`). Also we don't really "update" locations in the AST. We just get a brand new AST.